nixos/modules/services/matrix/synapse.nix

```diff
@@ -727,13 +784,124 @@ in {
               '';
             };
 
+            redis = lib.mkOption {
+              type = types.submodule {
+                freeformType = format.type;
+                options = {
+                  enabled = lib.mkOption {
+                    type = types.bool;
+                    default = false;
+                    description = lib.mdDoc ''
+                      Whether to use redis support
+                    '';
+                  };
+                };
+              };
+              default = { };
+              description = lib.mdDoc ''
+                Redis configuration for synapse.
+
+                See the
+                [upstream documentation](https://github.com/matrix-org/synapse/blob/v${pkgs.matrix-synapse-unwrapped.version}/usage/configuration/config_documentation.md#redis)
+                for available options.
+              '';
+            };
+          };
+        };
+      };
+
+      workers = lib.mkOption {
+        default = { };
+        description = lib.mdDoc ''
+          Options for configuring workers. See `services.matrix-synapse.workers.enable`
+          for a more detailed description.
+        '';
+        type = types.submodule {
+          options = {
+            enable = lib.mkOption {
+              type = types.bool;
+              default = false;
+              description = lib.mdDoc ''
+                Whether to enable matrix synapse workers.
+
+                ::: {.note}
+                  Enabling this will add a replication listener to the default
+                  value of `services.matrix-synapse.settings.listeners` and configure that
+                  listener as `services.matrix-synapse.settings.instance_map.main`.
+                  If you set either of those options, make sure to configure a replication
+                  listener yourself.
+
+                  A redis server is required for running workers. A local one can be enabled
+                  using `services.matrix-synapse.configureRedisLocally`.
```

Comment:
I think it might also be reasonable to explain that this is the place where to add worker-specific config that overrides stuff from the "shared" config (i.e. the declarations from `services.matrix-synapse.settings`). E.g. the shared config sets `enable_media_repo = false` whereas the media_repo worker's config sets it to `true` in here. This may be rather obvious when configuring synapse by hand because you specify multiple `-c` options with shared config and worker-specific config. This is however not transparent by this module (which is absolutely OK), but may cause to confusion (exhibit A: myself who was wondering how to do worker-specific config in here ;-) ).